install/0000_90_cluster-version-operator_02_servicemonitor.yaml

```diff
@@ -16,3 +16,49 @@ spec:
   selector:
     matchLabels:
       k8s-app: cluster-version-operator
+---
+apiVersion: monitoring.coreos.com/v1
+kind: PrometheusRule
+metadata:
+  labels:
+    k8s-app: cluster-version-operator
+  name: cluster-version-operator
+  namespace: openshift-cluster-version
+spec:
+  groups:
+  - name: cluster-version
+    rules:
+    - alert: ClusterVersionOperatorDown
+      annotations:
+        message: Cluster version operator has disappeared from Prometheus target discovery. Operator may be down or disabled, cluster will not be kept up to date and upgrades will not be possible.
+      expr: |
+        absent(up{job="cluster-version-operator"} == 1)
+      for: 15m
+      labels:
+        severity: critical
+  - name: cluster-operators
+    rules:
+    - alert: ClusterOperatorDown
+      annotations:
+        message: Cluster operator {{ "{{ $labels.name }}" }} has not been available for 10 mins. Operator may be down or disabled, cluster will not be kept up to date and upgrades will not be possible.
+      expr: |
+        cluster_operator_up{job="cluster-version-operator"} == 0
```

Comment:
How is this metric determined? And I assume this metric is somehow synthetically produced by the cluster-version-operator?

Reply:
> How is this metric determined? And I assume this metric is somehow synthetically produced by the cluster-version-operator?

what do we gain from knowing how this is generated?

it looks at clusteroperator and reports 1 when `available & ! degraded` else 0

https://github.com/openshift/cluster-version-operator/blob/e47c7784427479b738a596f7f633e2448ddf2dac/pkg/cvo/metrics.go#L262-L270

Reply:
There's probably a whole here where a rolling outage across operators could trigger the alert without having a single operator down for the full threshold.  For example:

1. Scrape 1, operator A is down.
2. Scrape 2, operator B is down; A is back up.
3. Scrape 3, operator C is down; B is back up.
4. Alert fires with "there has been at least one down operator for the last three scrapes (over 10m)"

I'm not sure how to adjust this query to avoid that though, although it plays into the "many operators behind one alert" point you raise [here][1].

[1]: https://github.com/openshift/cluster-version-operator/pull/232/files#r315432978

Reply:
These are metrics that should be exposed by the operators themselves and not by the CVO. That would remove the indirections, and we can make alerts concrete and actionable. As they are, these alerts, except for the very first one are not actionable, nor do they follow the best practice of metrics coming from the origin.

Reply:
See my other comment, but the metrics are part of the product.  An individual operator doesn’t report whether other operators see the result.  We *must* see the core loop view like this.

Reply:
I understand what you mean by these metrics describing the "advancing of the CVO", and I'm not arguing that it's not important, however, it is not clear to an SRE (even our SRE I've talked to them) if they are paged with this alert today what an actionable thing to troubleshoot is, not even what's broken. As is, this is at most actionable for our own engineering. As I said in the other comment, I want us to have a plan for how we will get to a state where this _is_ clear and it _is_ clear what about the product is not working when it's not. This needs to be top priority for the next release(s).